benches/codecs/character_delimited_bytes.rs

```diff
@@ -0,0 +1,90 @@
+};
+use std::{fmt, time::Duration};
+use tokio_util::codec::Decoder;
+use vector::codecs::{self, BytesDeserializer, CharacterDelimitedDecoder};
+
+#[derive(Debug)]
+struct Param {
+    slug: &'static str,
+    input: BytesMut,
+    max_length: Option<usize>,
+}
+
+impl fmt::Display for Param {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.slug)
+    }
+}
+
+fn decoding(c: &mut Criterion) {
+    let mut group: BenchmarkGroup<WallTime> =
+        c.benchmark_group("vector::codecs::decoding::Decoder");
+    group.sampling_mode(SamplingMode::Auto);
+
+    for param in &[
+        Param {
+            slug: "no_max",
+            input: BytesMut::from(include_str!("moby_dick.txt")),
+            max_length: None,
+        },
+        Param {
+            slug: "small_max",
+            input: BytesMut::from(include_str!("moby_dick.txt")),
+            max_length: Some(10),
+        },
+    ] {
+        group.throughput(Throughput::Bytes(param.input.len() as u64));
+        group.bench_with_input(
+            BenchmarkId::new("character_delimited_bytes", param),
+            &param,
+            |b, param| {
+                b.iter_batched(
+                    || {
+                        let framer = Box::new(
+                            param
+                                .max_length
+                                .map(|ml| CharacterDelimitedDecoder::new_with_max_length('a', ml))
+                                .unwrap_or(CharacterDelimitedDecoder::new('a')),
+                        );
+                        let deserializer = Box::new(BytesDeserializer::new());
+                        let decoder = codecs::Decoder::new(framer, deserializer);
+
+                        (Box::new(decoder), param.input.clone())
+                    },
+                    |(mut decoder, mut input)| loop {
+                        match decoder.decode_eof(&mut input) {
+                            Ok(Some(_)) => continue,
+                            Ok(None) => break,
+                            Err(_) => {
+                                unreachable!()
+                            }
+                        }
+                    },
+                    BatchSize::SmallInput,
+                )
+            },
+        );
+    }
+}
+
+criterion_group!(
+    name = benches;
+    config = Criterion::default()
+        .warm_up_time(Duration::from_secs(5))
+        .measurement_time(Duration::from_secs(120))
+        // degree of noise to ignore in measurements, here 1%
+        .noise_threshold(0.01)
+        // likelihood of noise registering as difference, here 5%
+        .significance_level(0.05)
+        // likelihood of capturing the true runtime, here 95%
+        .confidence_level(0.95)
+        // total number of bootstrap resamples, higher is less noisy but slower
+        .nresamples(100_000)
+        // total samples to collect within the set measurement time
+        .sample_size(150);
```

Comment:
Are those parameters something we (need to) tweak for every test, or could we move some of those to a shared config?

Reply:
They're contingent on the runtime behavior of the thing being benchmarked. We could potentially make confidence level a global configuration value. But I would also caution that the micro-benchmarks don't have a significant value in regression detection, so I hesitate to spend too much time making configuration frameworks or what not for them.